File: backend/models/Lift.js
// ============================================
// File: backend/models/Lift.js
// LIFT MANAGEMENT MODEL
// ============================================

const mongoose = require('mongoose');

const liftSchema = new mongoose.Schema({
  liftId: {
    type: String,
    required: true,
    unique: true,
  },
  blockId: {
    type: String,
    required: true,
    enum: ['BLOCK-A', 'BLOCK-B', 'BLOCK-C', 'BLOCK-D'],
  },
  liftNumber: {
    type: Number,
    required: true,
    enum: [1, 2],
  },
  status: {
    type: String,
    enum: ['available', 'occupied', 'in_transit', 'maintenance'],
    default: 'available',
  },
  currentBookingId: {
    type: String,
    default: null,
  },
  currentVehicleNumber: {
    type: String,
    default: null,
  },
  assignedAt: {
    type: Date,
    default: null,
  },
  releasedAt: {
    type: Date,
    default: null,
  },
  sensorStatus: {
    type: Boolean,
    default: false, // false = no car, true = car present
  },
  floor: {
    type: String,
    default: 'Ground',
  },
  lastActivity: {
    type: Date,
    default: Date.now,
  },
}, {
  timestamps: true,
});

// Index for faster queries
liftSchema.index({ blockId: 1, status: 1 });
liftSchema.index({ liftId: 1 });

const Lift = mongoose.model('Lift', liftSchema);

module.exports = Lift;

File: lib/models/lift.dart
class Lift {
  final String id;
  final String liftId;
  final String blockId;
  final int liftNumber;
  final String status;
  final String? currentBookingId;
  final String? currentVehicleNumber;
  final DateTime? assignedAt;
  final DateTime? releasedAt;
  final bool sensorStatus;
  final String floor;
  final DateTime lastActivity;

  Lift({
    required this.id,
    required this.liftId,
    required this.blockId,
    required this.liftNumber,
    required this.status,
    this.currentBookingId,
    this.currentVehicleNumber,
    this.assignedAt,
    this.releasedAt,
    required this.sensorStatus,
    required this.floor,
    required this.lastActivity,
  });

  factory Lift.fromJson(Map<String, dynamic> json) {
    print('🔍 Parsing Lift from JSON: $json');

    try {
      return Lift(
        id: json['_id']?.toString() ?? '',
        liftId: json['liftId']?.toString() ?? '',
        blockId: json['blockId']?.toString() ?? '',
        liftNumber: json['liftNumber'] is int
            ? json['liftNumber']
            : int.tryParse(json['liftNumber']?.toString() ?? '0') ?? 0,
        status: json['status']?.toString() ?? 'available',
        currentBookingId: json['currentBookingId']?.toString(),
        currentVehicleNumber: json['currentVehicleNumber']?.toString(),
        assignedAt: json['assignedAt'] != null
            ? DateTime.tryParse(json['assignedAt'].toString())
            : null,
        releasedAt: json['releasedAt'] != null
            ? DateTime.tryParse(json['releasedAt'].toString())
            : null,
        sensorStatus: json['sensorStatus'] == true,
        floor: json['floor']?.toString() ?? 'Ground',
        lastActivity: json['lastActivity'] != null
            ? DateTime.tryParse(json['lastActivity'].toString()) ??
                  DateTime.now()
            : DateTime.now(),
      );
    } catch (e) {
      print('❌ Error parsing Lift: $e');
      rethrow;
    }
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'liftId': liftId,
      'blockId': blockId,
      'liftNumber': liftNumber,
      'status': status,
      'currentBookingId': currentBookingId,
      'currentVehicleNumber': currentVehicleNumber,
      'assignedAt': assignedAt?.toIso8601String(),
      'releasedAt': releasedAt?.toIso8601String(),
      'sensorStatus': sensorStatus,
      'floor': floor,
      'lastActivity': lastActivity.toIso8601String(),
    };
  }

  bool get isAvailable => status == 'available';
  bool get isOccupied => status == 'occupied' || status == 'in_transit';
}
